File: hw2/LSclassify.py
#!/usr/bin/env python3
"""
LSClassify.py

Least squares clasification.
"""

import numpy
import scipy.linalg

def ls_classify(Xtrain, y, λ):
    """

    alpha, beta = ls_classify(Xtrain, y, λ)

    INPUT

    Xtrain -> training data 
    y   -> class labels for training data
    λ -> lambda value to use (`lambda` is a python builtin, haha)

    OUTPUT

    α -> scalar (offset of the separating hyperplane)
    β -> vector ('slope' of the separating hyperplane)
    
    a python implementation of LSclassify.m
    least squares classification given the training set Xtrain
    returns the parameters for the classifying hyperplane

    steps:

    (1) compute mean value centering
    (2) compute the SVD
    (3) compute alpha, beta

    """
    x_bar = Xtrain.mean(axis=0)  # average of each pixel (r,g,b); x_bar.size -> 3
    y_bar = y.mean() # average class label 
    
    # do mean centering (note: broadcasting)
    X_tilde = Xtrain - x_bar
    y_tilde = y - y_bar
    
    # full_matrices=False is the equivalent of MATLAB's "econ"
    # remember V is transposed of what it is in matlab
    U, S, V = scipy.linalg.svd(X_tilde, full_matrices=False)

    # c values
    C = S / (S*S + λ)

    # google "PEP 465" :)
    # check U.T and not U 
    β = numpy.dot(V.T, (C * numpy.dot(U.T, y_tilde)))
    α = y_bar - numpy.dot(x_bar, β)

    return α, β

if __name__ == "__main__":
    pass

File: matlab/util/LSClassify.m
function [alpha,beta] = LSClassify( Xtrain , y , lambda)
    %{
    Perform a least squares classification.

    author: Luke Wukmer
    Written for MATH 579, Spring 2015

    INPUT

    Xtrain: an mxn matrix. each of the m rows represents a data point in the
        system. this is the training data.
    y:      an mx1 vector, whose ith component is the 'class label' (0 or 1)
        of the ith data point (corresponding to the mth row of Xtrain)
    lambda: a parameter.

    OUTPUT:
    The components of the separating hyperplane f(x) = alpha + <beta, x>
    Then a data point x_ will be class 1 when f(x_) > .5, else class 0.

    alpha:  a scalar
    beta:   an nx1 vector
    %}

    % find the mean along columns of Xtrain. In other words, x_bar
    % is the componentwise average of all training points and has shape 1xn 
    xbar = mean(Xtrain, 1);

    % similarly, y_bar is the average class label. a scalar.
    ybar = mean(y);

    % this broadcasts the 1xn shape across Xtrain with the operation 'minus'
    Xtilde = bsxfun(@minus, Xtrain, xbar);

    % but broadcasting works normally like this
    ytilde = y - ybar;

    % SVD decompose the mean centered system
    [u, s, v] = svd(Xtilde, 'econ');

    % c values
    s_values = diag(s,0);
    c = s_values ./ (s_values.*s_values + lambda);

    % alpha and beta
    beta = v * (c .* (u'*ytilde));
    alpha = ybar - xbar*beta;